File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    private ListNode merge(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode();
        ListNode curr = dummy;

        while (list1 != null && list2 != null) {
            if (list1.val <= list2.val) {
                curr.next = list1;
                list1 = list1.next;
            } else {
                curr.next = list2;
                list2 = list2.next;
            }
            curr = curr.next;
        }

        while (list1 != null) {
            curr.next = list1;
            list1 = list1.next;
            curr = curr.next;
        }
        while (list2 != null) {
            curr.next = list2;
            list2 = list2.next;
            curr = curr.next;
        }
        return dummy.next;
    }

    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null;
        ListNode list = lists[0];
        for (int i = 1; i < lists.length; i++) {
            list = merge(list, lists[i]);
        }
        return list;
    }
}

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None

        def merge(list1, list2):
            dummy = ListNode()
            curr = dummy

            i = list1
            j = list2

            while i and j:
                if i.val < j.val:
                    curr.next = i
                    i = i.next
                else:
                    curr.next = j
                    j = j.next

                curr = curr.next

            while i:
                curr.next = i
                i = i.next
                curr = curr.next

            while j:
                curr.next = j
                j = j.next
                curr = curr.next

            return dummy.next

        list1 = lists[0]
        for i in range(1, len(lists)):
            list1 = merge(list1, lists[i])

        return list1